### PR TITLE
add preference feedback component api

### PR DESCRIPTION
**Pull request by moririn2528**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

### CODE REVIEW

File: optuna_dashboard/_preference_setting.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            "artifact_key": artifact_key,
+        },
+    )
+
+
+def register_preference_feedback_component_type(
+    study: PreferentialStudy,
+    component_type: OUTPUT_COMPONENT_TYPE,
+    artifact_key: str = "",
+) -> None:
```
Comment (contramundum53) [inline optuna_dashboard/_preference_setting.py:38]: If the user specifies `component_type="Artifact"` without setting the `artifact_key`, the current behavior is to set an empty string to `artifact_key`. As a result, the artifact will not be shown, and the user may not know what's wrong.

How about setting the default argument of `artifact_key` to `None` and perform a check here?

File: optuna_dashboard/_preference_setting.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+from optuna.storages import BaseStorage
+
+from .preferential._study import PreferentialStudy
+
+
+if TYPE_CHECKING:
+    from typing import Literal
+
+    OUTPUT_COMPONENT_TYPE = Literal["Note", "Artifact"]
```
Comment (contramundum53) [inline optuna_dashboard/_preference_setting.py:13]: How about using lower-case, to be consistent with "minimum"/"maximum" used in optuna?

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Comment by contramundum53** [comment]
Please format the code by 
```
black .
isort .
```

File: optuna_dashboard/_app.py [hunk 3]
```diff
@@ -309,6 +310,26 @@ def post_preference(study_id: int) -> dict[str, Any]:
...
+        if component_type not in ["note", "artifact"]:
+            response.status = 400
+            return {"reason": "component_type must be either 'Note' or 'Artifact'."}
+
+        _register_preference_feedback_component_type(
+            study_id=study_id,
+            storage=storage,
+            component_type=component_type,
+            artifact_key=artifact_key,
+        )
```
Comment (c-bata) [inline optuna_dashboard/_app.py:331]: I think you need to set `response.status = 204` to make tests pass.

File: optuna_dashboard/_app.py [hunk 4]
```diff
@@ -309,6 +310,26 @@ def post_preference(study_id: int) -> dict[str, Any]:
         response.status = 204
         return {}
 
+    @app.put("/api/studies/<study_id:int>/preference_feedback_component_type")
+    @json_api_view
+    def put_component(study_id: int) -> dict[str, Any]:
```
Comment (c-bata) [inline optuna_dashboard/_app.py:315]: Could you update the view function name to elaborate the role of this api endpoint?

File: optuna_dashboard/_serializer.py [hunk 5]
```diff
@@ -163,6 +164,10 @@ def serialize_study_detail(
     form_widgets = get_form_widgets_json(system_attrs)
     if form_widgets:
         serialized["form_widgets"] = form_widgets
+    if serialized["is_preferential"]:
+        serialized["feedback_component_type"] = system_attrs.get(
+            _SYSTEM_ATTR_FEEDBACK_COMPONENT, {}
+        )
     if serialized["is_preferential"]:
```
Comment (c-bata) [inline optuna_dashboard/_serializer.py:171]: I think you can merge this if-statement block.

**Review by contramundum53 — APPROVED** [review]
LGTM!

File: optuna_dashboard/_preference_setting.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    Args:
+        study:
+            The study to register the output component.
+        component_type:
+            The type of the output component.
+        artifact_key:
+            When the component_type is "Artifact",
+            this argument is used as the attribute key of the artifact.
+            Each trial displays the artifact whose id is the value of the attribute.
+    """
```
Comment (c-bata) [inline optuna_dashboard/_preference_setting.py:51]: How about updating the function name and docstring like below?

```suggestion
def register_preference_feedback_component(
    study: PreferentialStudy,
    component_type: OUTPUT_COMPONENT_TYPE,
    user_attr_artifact_key: str | None = None,
) -> None:
    """Register a preference feedback component to the study.
    
    With this feature, you can change the component, displayed on the
    human feedback pages. By default, the Markdown note (``component_type="note"``)
    is displayed.  If you specify ``component_type="user_attr"``, the viewer for the
    specified artifact file will be displayed.

    Args:
        study:
            The study to register the preference feedback component.
        component_type:
            The component type, displayed on the human feedback pages
            (default: ``note``).
        user_attr_artifact_key:
            This option is required when the ``component_type`` is ``"artifact``.
            The user attribute, which is specified this field, must contain the
            ``artifact`id you want to display on the human feedback page.
    """
```

**Review by c-bata — CHANGES_REQUESTED** [review]

**Review by c-bata — APPROVED** [review]
LGTM.

File: optuna_dashboard/_preference_setting.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    if artifact_key is not None:
+        value["artifact_key"] = artifact_key
+    storage.set_study_system_attr(
+        study_id=study_id,
+        key=_SYSTEM_ATTR_FEEDBACK_COMPONENT,
+        value=value,
+    )
+
+
+def register_preference_feedback_component_type(
```
Comment (c-bata) [inline optuna_dashboard/_preference_setting.py:35]:
```suggestion
def register_preference_feedback_component(
```

**Review by c-bata — COMMENTED** [review]
Now I noticed that we need to update the api reference below since this PR adds a public api. Could you update it?
https://optuna-dashboard--615.org.readthedocs.build/en/615/api.html#preferential-optimization

File: docs/api.rst [hunk 8]
```diff
@@ -44,6 +44,7 @@ Preferential Optimization
    optuna_dashboard.preferential.create_study
    optuna_dashboard.preferential.load_study
    optuna_dashboard.preferential.PreferentialStudy
+   optuna_dashboard._preference_setting.register_preference_feedback_component
```
Comment (c-bata) [inline docs/api.rst:47]: You need to export from the public module. Could you update the `__init__.py` as well?

```suggestion
   optuna_dashboard.register_preference_feedback_component
```

**Pull request merged**